src/tree.rs

```diff
@@ -400,6 +406,89 @@ impl<T: Value> Tree<T> {
             }
         }
     }
+
+    /// Exploit structural sharing between identical parts of the tree.
+    ///
+    /// This method traverses a fully-hashed tree and replaces identical subtrees with clones of
+    /// the first equal subtree. The result is a tree that shares memory for common subtrees, and
+    /// thus uses less memory overall.
+    ///
+    /// You MUST pass a fully-hashed tree to this function, or an `Error::IntraRebaseZeroHash`
+    /// error will be returned.
+    ///
+    /// Arguments are:
+    ///
+    /// - `orig`: The tree to rebase.
+    /// - `known_subtrees`: map from `(depth, tree_hash_root)` to `Arc<Node>`. This should be empty
+    ///    for the top-level call. The recursive calls fill it in. It can be discarded after the
+    ///    method returns.
+    /// - `current_depth`: The depth of the tree `orig`. This will be decremented as we recurse
+    ///    down the tree towards the leaves.
+    ///
+    /// Presently leaves are left untouched by this procedure, so it will only produce savings in
+    /// trees with equal internal nodes (i.e. equal subtrees with at least two leaves/packed leaves
+    /// under them).
+    ///
+    /// The input tree must be fully-hashed, and the result will also remain fully-hashed.
+    pub fn intra_rebase(
+        orig: &Arc<Self>,
+        known_subtrees: &mut HashMap<(usize, Hash256), Arc<Self>>,
+        current_depth: usize,
+    ) -> Result<IntraRebaseAction<Self>, Error> {
+        match &**orig {
+            Self::Leaf(_) | Self::PackedLeaf(_) | Self::Zero(_) => Ok(IntraRebaseAction::Noop),
+            Self::Node { hash, left, right } if current_depth > 0 => {
+                let hash = *hash.read();
+
+                // Tree must be fully hashed prior to intra-rebase.
+                if hash.is_zero() {
+                    return Err(Error::IntraRebaseZeroHash);
+                }
+
+                if let Some(known_subtree) = known_subtrees.get(&(current_depth, hash)) {
+                    // Node is already known from elsewhere in the tree. We can replace it without
+                    // looking at further subtrees.
+                    return Ok(IntraRebaseAction::Replace(known_subtree.clone()));
```

Comment:
A small further optimisation here would be to avoid doing a `Replace` if the `known_subtree` and `orig` are already ptr-equal, e.g. if we call `intra_rebase` twice in a row it should always be a `Noop` (we could write a test for this).